### PR TITLE
Node.js AWS Task #6 Solution

### PR DESCRIPTION
**Pull request by OHalahan**
**TASK 6.1 ✅**
**TASK 6.2 ✅**
**TASK 6.3 ✅**

**Integrated FE link**
https://d3ehhvbc9pzoid.cloudfront.net/admin/products

**email example**
image

**catalogBatchProcess -> SQS relation**
image

**CSV Example**
``` csv
title,description,price,count
New Hublot,some description,200500,3
```

## Evaluation criteria ✅

Reviewers should verify the lambda functions, SQS and SNS topic and subscription in PR.

1 - File **serverless.yml** contains configuration for **catalogBatchProcess** function ✅
2 - File **serverless.yml** contains policies to allow lambda *catalogBatchProcess** function to interact with SNS and SQS ✅
3 - File **serverless.yml** contains configuration for SQS **catalogItemsQueue** ✅
4 - File **serverless.yml** contains configuration for SNS Topic **createProductTopic** and email subscription ✅

## Additional (optional) tasks ✅
+1 - **catalogBatchProcess** lambda is covered by unit tests ✅
image

+1 - set a Filter Policy for SNS **createProductTopic** in **serverless.yml** (Create an additional email subscription and distribute messages to different emails depending on the filter for any product attribute) ✅
> notification to additional email will be sent if price is less than $20k

### CODE REVIEW

**Comment by kotmatpockuh** [comment]
The best PR description I ever saw during this course 🥳

File: import-service/handlers/importFileParser.ts [hunk 1]
```diff
@@ -1,13 +1,15 @@
...
+import {SQS} from 'aws-sdk';
 import csv from 'csv-parser';
 import util from 'util';
 import stream from 'stream';
 
 import {AWS_REGION, BUCKET_NAME, SOURCE_DIR, TARGET_DIR} from '../constants';
 
 export const importFileParser = async (event: S3Event) => {
+  const sqs = new AWS.SQS({region: AWS_REGION});
   const s3 = new AWS.S3({region: AWS_REGION});
```
Comment (kotmatpockuh) [inline import-service/handlers/importFileParser.ts:13]: `{region: AWS_REGION}` - duplication should be refactored

Comment (OHalahan) [reply import-service/handlers/importFileParser.ts]: Fair point. Optimized.
Thank you :)

File: import-service/handlers/importFileParser.ts [hunk 2]
```diff
@@ -48,14 +59,33 @@ export const importFileParser = async (event: S3Event) => {
   return;
 };
 
-class LogEachRow extends stream.Transform {
-  constructor() {
+class PassToSQS extends stream.Transform {
```
Comment (SkorykV) [inline import-service/handlers/importFileParser.ts:62]: heh, good idea tosend sqs message in a transform step. Liked it!

**Review by SkorykV — APPROVED** [review]

File: import-service/handlers/importFileParser.ts [hunk 3]
```diff
@@ -20,7 +24,14 @@ export const importFileParser = async (event: S3Event) => {
...
       }).createReadStream();
 
-      await pipeline(s3Stream, csv(), new LogEachRow());
+      await pipeline(
+        s3Stream,
+        csv({
+          // cast numeric values to Number
+          mapValues: ({value}) => isNaN(value) ? value : Number(value)
+        }),
+        new PassToSQS(sqs)
```
Comment (yankouskia) [inline import-service/handlers/importFileParser.ts:33]: 👍

File: import-service/handlers/importFileParser.ts (hunk 2 above)
Comment (OHalahan) [reply import-service/handlers/importFileParser.ts:62]: thank you :)

**Pull request merged**